src/librbd/image/ListWatchersRequest.cc

```diff
@@ -0,0 +1,176 @@
+    finish(r);
+    return;
+  }
+
+  get_mirror_image();
+}
+
+template<typename I>
+void ListWatchersRequest<I>::get_mirror_image() {
+  ldout(m_cct, 20) << dendl;
+  if ((m_object_watchers.empty()) ||
+      ((m_image_ctx.features & RBD_FEATURE_JOURNALING) == 0)) {
+    finish(0);
+    return;
+  }
+
+  librados::ObjectReadOperation op;
+  cls_client::mirror_image_get_start(&op, m_image_ctx.id);
+
+  using klass = ListWatchersRequest<I>;
+  librados::AioCompletion *comp =
+    create_rados_callback<klass, &klass::handle_get_mirror_image>(this);
+  m_out_bl.clear();
+  int r = m_image_ctx.md_ctx.aio_operate(RBD_MIRRORING, comp, &op, &m_out_bl);
+  assert(r == 0);
+  comp->release();
+}
+
+template<typename I>
+void ListWatchersRequest<I>::handle_get_mirror_image(int r) {
+  ldout(m_cct, 20) << "r=" << r << dendl;
+
+  if (r == -ENOENT || r == -EOPNOTSUPP) {
+    finish(0);
+    return;
+  } else if (r < 0) {
+    ldout(m_cct, 1) << "error retrieving mirror image: " << cpp_strerror(r)
+                    << dendl;
+  }
+
+  list_mirror_watchers();
+}
+
+template<typename I>
+void ListWatchersRequest<I>::list_mirror_watchers() {
+  ldout(m_cct, 20) << dendl;
+
+  librados::ObjectReadOperation op;
+  op.list_watchers(&m_mirror_watchers, &m_ret_val);
+
+  using klass = ListWatchersRequest<I>;
+  librados::AioCompletion *rados_completion =
+    create_rados_callback<klass, &klass::handle_list_mirror_watchers>(this);
+  m_out_bl.clear();
+  int r = m_image_ctx.md_ctx.aio_operate(RBD_MIRRORING, rados_completion,
+                                         &op, &m_out_bl);
+  assert(r == 0);
+  rados_completion->release();
+}
+
+template<typename I>
+void ListWatchersRequest<I>::handle_list_mirror_watchers(int r) {
+  ldout(m_cct, 20) << "r=" << r << dendl;
+
+  if (r == 0 && m_ret_val < 0) {
+    r = m_ret_val;
+  }
+  if (r < 0 && r != -ENOENT) {
+    ldout(m_cct, 1) << "error listing mirror watchers: " << cpp_strerror(r)
+                    << dendl;
+  }
+  finish(0);
+}
+
+template<typename I>
+void ListWatchersRequest<I>::finish(int r) {
+  ldout(m_cct, 20) << "r=" << r << dendl;
+
+  if (r == 0) {
+    m_watchers->clear();
+
+    if (m_object_watchers.size() > 0) {
+      RWLock::RLocker owner_locker(m_image_ctx.owner_lock);
+      uint64_t watch_handle = m_image_ctx.image_watcher != nullptr ?
+        m_image_ctx.image_watcher->get_watch_handle() : 0;
+
+      for (auto &w : m_object_watchers) {
+        auto it = std::find_if(m_mirror_watchers.begin(),
+                               m_mirror_watchers.end(),
+                               [w] (obj_watch_t &watcher) {
+                                 return (strncmp(w.addr, watcher.addr,
+                                                 sizeof(w.addr)) == 0);
+                               });
+        bool mirroring = (it != m_mirror_watchers.end());
+        bool me = (w.cookie == watch_handle); // XXXMG: does it work?
+        m_watchers->emplace_back(w.addr, w.watcher_id, w.cookie,
```

Comment:
Nit: perhaps just filter out yourself and mirroring watchers?

Reply:
Do you mean that `ListWatchersRequest` should return already filtered list of watchers (of obj_watch_t type, no need to add WatcherInfo then, or just return a number of watchers)?

I would like to have it generic. Not sure it would ever use for other purposes though, but for example, recently we have added list_watchers to our API. It  might be not too late to modify this function to also return "me" and "mirroring" info?

Or at least the request name should be something like `ListWatchersButMeAndMirrorRequest` then, in order one could understand what it did without looking to its code. But it looks a little awkward. May be `ListBlockingWatchersRequest`?

Reply:
Maybe just provide some optional filter flags to the constructor to filter out mirroring daemon clients and/or yourself?